devito/data/data.py

```diff
@@ -195,32 +196,63 @@ def __getitem__(self, glb_idx, comm_type):
                           modulo=(False,)*len(local_val.shape))
             it = np.nditer(owners, flags=['refs_ok', 'multi_index'])
             # Iterate over each element of data
-            while not it.finished:
-                index = it.multi_index
-                if rank == owners[index] and rank == send[index]:
-                    # Current index and destination index are on the same rank
-                    loc_ind = local_si[index]
-                    send_ind = local_si[global_si[index]]
-                    retval.data[send_ind] = local_val.data[loc_ind]
-                elif rank == owners[index]:
-                    # Current index is on this rank and hence need to send
-                    # the data to the appropriate rank
-                    loc_ind = local_si[index]
-                    send_rank = send[index]
-                    send_ind = global_si[index]
-                    send_val = local_val.data[loc_ind]
-                    reqs = comm.isend([send_ind, send_val], dest=send_rank)
-                    reqs.wait()
-                elif rank == send[index]:
-                    # Current rank is required to receive data from this index
-                    recval = comm.irecv(source=owners[index])
-                    local_dat = recval.wait()
-                    loc_ind = local_si[local_dat[0]]
-                    retval.data[loc_ind] = local_dat[1]
+            # TODO: These statements can potentially be fused?
```

Comment:
yeah I can see there's some overlap... also the while-if-elif-else structure is identical AFAICT. We might want to abstract it away into a separate routine which takes callbacks (the bodies each each if-branch) as input? or is it just painful over-engineering? 
in the future, we might want to add other gather-like operations -- scatter, broadcast, ...

Reply:
Yes, it is identical at the moment. It possibly should stay identical depending on what functionality we want to allow - but we can discuss that.

> We might want to abstract it away into a separate routine

Yes, we definitely want to do that - it will make the fusion much cleaner. I'm just keeping them separate for the time being until I get all required functionality working with the gather, after that I'll have a good idea re. the best way to fuse them.